### PR TITLE
[SPARK-7294][SQL] ADD BETWEEN

### PR DESCRIPTION
**Pull request by kaka1992**

### CODE REVIEW

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -1289,6 +1289,18 @@ def cast(self, dataType):
...
+    @ignore_unicode_prefix
+    def between(self, col1, col2):
+        """ A boolean expression that is evaluated to true if the value of this
+        expression is between the given columns.
+
+        >>> df[df.col1.between(col2, col3)].collect()
+        [Row(col1=5, col2=6, col3=8)]
+        """
+        #sc = SparkContext._active_spark_context
+        jc = self > col1 & self < col2
```
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: i think u need to add parenthesis, i.e. (self > col1) & (self < col2)

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 2]
```diff
@@ -296,6 +296,20 @@ class Column(protected[sql] val expr: Expression) extends Logging {
...
+   * @group java_expr_ops
+   */
+  def between(col1: String, col2: String): Column = between(Column(col1), Column(col2))
+
+  /**
+   * Between col1 and col2.
+   *
+   * @group java_expr_ops
+   */
+  def between(col1: Column, col2: Column): Column = And(GreaterThan(this.expr, col1.expr), LessThan(this.expr, col2.expr))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: you'd need to wrap this so the line fits in 100 chars.

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 3]
```diff
@@ -296,6 +296,20 @@ class Column(protected[sql] val expr: Expression) extends Logging {
   def eqNullSafe(other: Any): Column = this <=> other
 
   /**
+   * Between col1 and col2.
+   *
+   * @group java_expr_ops
+   */
+  def between(col1: String, col2: String): Column = between(Column(col1), Column(col2))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: maybe name the two parameters  lowerBound and upperBound

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala (hunk 2 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: Note that between is really >= and <=, i.e. inclusive.

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 4]
```diff
@@ -296,6 +296,20 @@ class Column(protected[sql] val expr: Expression) extends Logging {
...
+
+  /**
+   * Between col1 and col2.
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: "True if the current column is between the lower bound and upper bound, inclusive."

File: python/pyspark/sql/dataframe.py [hunk 5]
```diff
@@ -1289,6 +1289,17 @@ def cast(self, dataType):
...
+    @ignore_unicode_prefix
+    def between(self, lowerBound, upperBound):
+        """ A boolean expression that is evaluated to true if the value of this
+        expression is between the given columns.
+
+        >>> df[df.col1.between(lowerBound, upperBound)].collect()
+        [Row(col1=5, col2=6, col3=8)]
+        """
+        jc = (self >= lowerBound) & (self <= upperBound)
+        return Column(jc)
```
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: actually I think you no longer need to wrap it in Column, since it is already a Python column.

**Comment by rxin** [comment]
Jenkins, ok to test.

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 6]
```diff
@@ -296,6 +296,25 @@ class Column(protected[sql] val expr: Expression) extends Logging {
...
+    between(Column(lowerBound), Column(upperBound))
+  }
+
+  /**
+   * True if the current column is between the lower bound and upper bound, inclusive.
+   *
+   * @group java_expr_ops
+   */
+  def between(lowerBound: Column, upperBound: Column): Column = {
+    And(GreaterThanOrEqual(this.expr, lowerBound.expr),
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: actually - how about we change the implementation here to use Column directly, i.e.
```scala
(this >= lowerBound) && (this <= upperBound)
```

and the Python one simply calls the JVM function?

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 7]
```diff
@@ -208,6 +208,12 @@ class ColumnExpressionSuite extends QueryTest {
       testData2.collect().toSeq.filter(r => r.getInt(0) <= r.getInt(1)))
   }
 
+  test("between") {
+    checkAnswer(
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: it'd be easier to just inline the data here I think, you can do
```scala
val testData = Seq((0, 1, 2), (1, 2, 3), ...).toDF("a", "b", "c")
```

File: python/pyspark/sql/tests.py [hunk 8]
```diff
@@ -426,6 +426,12 @@ def test_rand_functions(self):
         for row in rndn:
             assert row[1] >= -4.0 and row[1] <= 4.0, "got: %s" % row[1]
 
+    def test_between_function(self):
+        df = self.sqlCtx.parallelize([Row(a=1, b=2, c=3), Row(a=2, b=1, c=3), Row(a=4, b=1, c=4)]).toDF()
```
Comment (rxin) [inline python/pyspark/sql/tests.py]: i think this line would fail pep8 because it is over 100 chars

**Comment by kaka1992** [comment]
please test this pr.

**Comment by rxin** [comment]
Did you address my comments?

**Comment by kaka1992** [comment]
@rxin Please test this RR.

File: python/pyspark/sql/tests.py [hunk 9]
```diff
@@ -438,6 +438,14 @@ def test_rand_functions(self):
         for row in rndn:
             assert row[1] >= -4.0 and row[1] <= 4.0, "got: %s" % row[1]
 
+    def test_between_function(self):
+        df = self.sqlCtx.parallelize([Row(a=1, b=2, c=3),
+                                      Row(a=2, b=1, c=3),
+                                      Row(a=4, b=1, c=4)]).toDF()
+        self.assertEqual([False, True, True],
+                     df.select(df.a.between(df.b, df.c)).collect())
+
```
Comment (rxin) [inline python/pyspark/sql/tests.py]: remove one less blank line.

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 10]
```diff
@@ -296,6 +296,24 @@ class Column(protected[sql] val expr: Expression) extends Logging {
   def eqNullSafe(other: Any): Column = this <=> other
 
   /**
+   * True if the current column is between the lower bound and upper bound, inclusive.
+   *
+   * @group java_expr_ops
+   */
+  def between(lowerBound: String, upperBound: String): Column = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: i think you can remove this function, and change the signature of the one below to take "Any".

**Comment by kaka1992** [comment]
@rxin Please test this PR.

**Comment by rxin** [comment]
Jenkins, ok to test.

**Comment by kaka1992** [comment]
Oro failure. Please retest it.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by kaka1992** [comment]
Please retest this =_=. @rxin

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Thanks. I've merged this.

**Pull request closed without merging**